첫 회원가입 화면 만들기

GangWoon opened the pull request:
스토리 보드를 적극적으로 사용하려고 노력했습니다.
텍스트 필드를 구분할 때 restoration id를 사용한게 잘 사용한건지 궁금합니다. 어차피 iboutlet을 선언해서 그냥 textField자체를 비교하는 방법도 있었습니다. 그리고 textField를 넘어갈때 next버튼을 누르면 버그가 없었지만 키보드로 tab을 눌렀을때 다른 텍스트 필드를 건드리는 부분을 확인했습니다. 위 상황도 명시적으로 ibout을 비교하는게 맞는지 아닌지가 궁금합니다...
아직 규모가 작아서 그런지 아니면 못 나눈건지 뷰에서 모든 작업을 처리하고 있었습니다. 만약에 뷰의 datasource를 다른 모델 객체로 넘겨준다면 스토리보드에서 어떻게 해야할까요?? 아니면 코드로 생성해야되는데 스토리 보드와 코드를 혼합해서 사용하면 느낌이 너무 이상하다고 생각했습니다. 
다음 화면으로 넘어갈때 버튼 엑션이 뷰에 있을텐데 이런 경우에는 뷰가 다음 뷰컨을 알고 있기때문에 mvc가 붕괴되지 않나요??

---
다른 팀원이랑 애기를 통해서 모델을 나누는 게 맞다고 생각이 들어서 수정했습니다.
그에따른 전반적으로 코드를 수정했습니다.


아래 사진은 tab으로 넘겼을때 발생하는 상황입니다.
<img width="545" alt="Screen Shot 2020-03-25 at 14 25 44" src="https://user-images.githubusercontent.com/48466830/77504737-84cde780-6ea4-11ea-86e5-d23960d103e3.png">

GangWoon committed:
ui: 첫번째 화면 UI 작성

GangWoon committed:
add: SignUpCustomView class 생성

mvc 패턴을 적용하기 위해서 view class를 생성했습니다.
IBOutletCollection을 사용하지 않은 이유는 아웃렛 자체를 호출할 일이 빈번하기때문에 인덱스로 접근하는 것 보다 선언해두고 사용하는게 더 적합하다고 생각을 했습니다.

GangWoon committed:
add: String extension

정규 표현식을 사용하기 위해서 string 구조체를 확장 시켰습니다.

GangWoon committed:
ui: 오토레이아웃 수정, textField restoration id 변경

전반적인 오토레이아웃 수정했습니다.
textField에 접근하기 위해서 restoration id를 사용하기로 결정했습니다.

GangWoon committed:
ui: 키보드 리턴키 변경, 암호화 설정

키보드 리턴키를 next와 done으로 변경했습니다.
비밀번호를 입력하는 부분에선 ** 표시가 되도록 변경했습니다.

GangWoon committed:
feat: TextField에 정규식 적용

각 텍스트 필드마다 다른 정규식을 사용해서 결과를 표시하도록 설정했습니다.
textField restoration id를 스트링에 저장할까 텍스트 필드에 저장할까 커스텀 클래스에 저장할까 고민했습니다.
처음 코드를 작성했을때는 shouldReturn, endEditing 모두 버튼 활성화를 검사했지만 둘중 하나만 불리는 게 아닌 둘다 불리는 걸 확인해서 한 곳에만 작성했습니다.
textField를 검사하는 방식이 올바른지 궁금합니다.

GangWoon committed:
add: MVC 객체 생성

GangWoon committed:
refactor: SignUpView  클래스 변경

custom이란 이름이 너무 이질 적으로 느껴서 변경했습니다.

GangWoon committed:
add: RegularExpression 프로토콜 생성

model 객체가 뷰객체의 이벤트를 처리하기 위해서 만들었습니다.
스토리 보드로만 작업해야된다고 생각했었는데 다른 사람과 이야기하면서 굳이 그럴 필요가 없다고 느껴서 분리했습니다.

GangWoon committed:
chore: View 객체와 Model 객체 바인딩

GangWoon committed:
chore: Model 객체 수정

뷰에서 이벤트가 발생했을때 모델에서 처리하고 반영할 수 있도록 작성했습니다. 다시 돌아갈 방법은 completion Closure를 사용했습니다.
정규식을 적용하지 않고 view에서 판단 가능한 부분은 모델에서 처리하지 않고 view에서 처리하도록 했습니다.

GangWoon committed:
refactor: 컨밴션 유지

godrm commented on ios/SignUp/SignUp/Protocol/RegularExpression.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+protocol RegularExpression: NSObject {
+    func regex(_ text: String, kinds: String, completion: (Bool, String, String) -> Void)
```
정규표현식이라는 것은 알겠는데, 어떤 객체들이 어떤 목적으로 써야 하는걸까요? 
프로토콜 이름이나 메소드 시그니처를 봐서는 구분할 수가 없네요. 기능 자체보다 프로토콜의 의도를 표현하면 좋겠습니다.

godrm commented on ios/SignUp/SignUp/SignUp/View/SignUpView.swift:
```diff
@@ -0,0 +1,121 @@
+    private func activateNextButton() {
+        let result = idetnfierStatusLabel.textColor == .systemGreen
+            && passwordStatusLabel.textColor == .systemGreen
+            && confirmPasswordStatusLabel.textColor == .systemGreen
```
뷰의 컬러값을 기준으로 비교하는 것은 좋은 것 같지 않네요. 디자인 요구사항이 로직과 분리되는 게 좋습니다. 
뷰에서 입력한 값으로 판단하는 로직을 가진 객체나 값을 비교하는 게 좋습니다. 

메소드 이름이 activateNextButton() 이면 activate처리를 해야 하는데 내부에서 다시 판단을 하고 있습니다.
이름과 일치하지 않는 동작이네요. 
이렇게 판단하는 것과 표시하는 것을 분리해서 생각하세요. 
판단하는 것은 로직이라 뷰 컨트롤러나 모델에서 하고, 출력하고 표시하는 것은 뷰에서 할 일입니다. 
그 부분이 안되고 있습니다.

godrm commented on ios/SignUp/SignUp/SignUp/View/SignUpView.swift:
```diff
@@ -0,0 +1,121 @@
+    private let identifierTextFieldKey: String = "identifier"
+    private let passwordTextFieldKey: String = "password"
+    private let confirmPasswordTextFieldKey: String = "confirmPassword"
+    private let nameTextFieldKey: String = "name"
```
TextFieldKey가 문자열로 왜 필요한지 모르겠습니다.

godrm commented on ios/SignUp/SignUp/SignUp/View/SignUpView.swift:
```diff
@@ -0,0 +1,121 @@
+    }
+    
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        switch textField.restorationIdentifier {
```
restorationIdentifier 역할이 무엇이고 어떤 용도로 써야하는 걸까요?
IBOutlet 객체 자체를 비교하면 되는게 아닐까요?

godrm commented on ios/SignUp/SignUp/SignUp/View/SignUpView.swift:
```diff
@@ -0,0 +1,121 @@
+        guard let text = textField.text else { return }
+        switch textField.restorationIdentifier {
+        case identifierTextFieldKey:
+            datasoure?.regex(text, kinds: "identifier") { result, fail, sucess  in
```
`데이터 소스`는 뷰의 화면에 출력하기 위한 데이터를 갖고 있는 소스를 의미하는 게 아닌가요?

모든 textfield 이벤트를 여기서 모아서 한꺼번에 검증을 해야만 할까요? 
이런 단위는 뷰와 뷰 컨트롤러 관계를 분리하는 게 좋습니다. 
뷰는 입력 또는 출력만 담당하고, 그 사이에서 비교하거나 하는 것은 뷰 컨트롤러가 하는 거죠.
그래서 Container ViewController 라는 개념을 사용합니다.

뷰 객체에 왜 데이터소스와 델리게이트가 존재하는 걸까요?
뷰 로직을 가볍게 만들고, 비교하고, 계산하는 로직을 분리하기 위한 겁니다.
지금은 뷰가 모든 것을 다하고 있네요. 
switch-case 나 if 구문으로 분리하기 보다는 다형성으로 각 객체가 동작하도록 하는게 좋습니다.

GangWoon replied:
피드백 반영했습니다!! 
리뷰 부탁드리기 죄송하지만 제 나름대로 노력해서 코드를 개편했습니다!

godrm reviewed (commented):
질문한 내용이 왜 그런지 코드를 보고 이해했습니다.
뷰에서 모든 것을 다하려고 하기 때문에 그렇습니다.
뷰 역할과 뷰 컨트롤러, 데이터 소스, 델리게이터 역할을 다시 한 번 고민해보세요. 
이거 매우 중요한 경험이 될 겁니다.

GangWoon committed:
add: 새로운 파일들 추가

GangWoon committed:
chore: Notification.Name 전역 변수 생성

GangWoon committed:
chore: Typealias 추가

튜플을 사용할 때 그냥 사용하면 의도를 명확하게 파악하기 힘들어서 새로운 타입을 만들어봤습니다.

GangWoon committed:
refactor: 비교에 대한 작업을 매니저에서 처리

뷰에서 처리했었던 간단한 작업까지도 매니저에서 처리하도록 변경했습니다.

GangWoon committed:
refactor: 프로토콜 이름 변경

SignUpViewModel로 변경했습니다.
사실 이번 프로젝트를 진행하면서 생각은 정말 많이했지만 함수, 변수, 클래스, 프로토콜 이름에 대해서 많이 고민하지 못했습니다.
바르게 작성한지 모르겠습니다.

GangWoon committed:
refactor: Delegate 객체 생성

한곳에서 모든걸 처리했었는데 작업을 나누기 위해서 각각의 delegate 객체를 생성했습니다.

GangWoon committed:
feat: ConfirmPasswordTextField Delegate 생성

passwordTextField의 문자열을 받아오고 비교하는 현재(ConfrimTextField) 문자열과 비교할 수 있는 객체를 생성했습니다.

GangWoon committed:
feat: IdentifierTextField Delegate 생성

identifierTextField의 문자열을 받아와서 유효성 검사를 하는 객체를 생성했습니다.
여기서 고민했던 점이 SignUpViewModel 프로토콜 타입 변수를 갖을건지 아니면 매니저로 노티피케이션을 보낼지 고민했었습니다.
둘다 이상하다는 느낌을 많이 받았지만 어떻게 해결할 방법을 몰라서 viewModel을 속성으로 갖는 방법을 선택했습니다.

GangWoon committed:
feat: NameTextField Delegate 생성

텍스트 필드의 값이 있는지 없는지 확인하는 객체를 생성했습니다.

GangWoon committed:
feat: PasswordTextField Delegate 생성

passwordTextField의 문자열을 받아와서 유효성 검사를 하는 객체를 생성했습니다.

GangWoon committed:
refactor: SignUpView 클래스 수정

기본적으로 계산하는 로직을 전부 분리했습니다.
수 많은 노티피케이션을 이용해서 자료를 넘겨주고 정보를 받도록 작성했습니다.
옵저버를 디이닛을 전부다 작성하지 못했습니다.
시간이 이렇게 촉박할지 계산하지 못해서 발생한 실수입니다.

GangWoon committed:
chore: 스토리보드, 파일 커밋

GangWoon commented:
피드백을 반영하려고 노력했지만, 코드 설계를 하면서 "이렇게 해도 괜찮은가"라는 생각이 많았습니다.
시간이 너무 촉박해서 변수명, 함수명을 생각하는 시간이 부족했고, 우선 기능 구현에 생각을 두다보니 코드를 조금 더 검토하고 올렸어야 했지만 그러지 못해서 죄송합니다

GangWoon commented:
![Paper 다이어리 95](https://user-images.githubusercontent.com/48466830/77729281-9ef0fc80-7041-11ea-9270-ee074e39ac9f.png)
